kv/dlock/dlock.go

```diff
@@ -0,0 +1,194 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dlock
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/tikv/oracle"
+	"github.com/pingcap/tidb/terror"
+	"github.com/twinj/uuid"
+)
+
+const (
+	leaseInterval time.Duration = 3 * time.Second
+)
+
+var (
+	// ErrAlreadyLocked is an error to indicate that the resource already locked by others.
+	ErrAlreadyLocked = errors.New("already locked")
+)
+
+// DLock represents an distributed lock that can be locked and unlocked.
+type DLock struct {
+	name         string
+	s            kv.Storage
+	unlockSignal chan struct{}
+	LockMeta
+}
+
+// LockMeta is the meta information that stored to the KV storage.
+type LockMeta struct {
+	ID          string    `json:"uuid"`
+	LastUpdated time.Time `json:"last_updated"`
+}
+
+// New returns a DLock object.
+func New(s kv.Storage, name string) *DLock {
+	return &DLock{
+		name:         name,
+		s:            s,
+		unlockSignal: make(chan struct{}),
+		LockMeta: LockMeta{
+			ID: uuid.NewV4().String(),
+		},
+	}
+}
+
+// TryLock tries to acquire the lock, returns error on fail.
+func (l *DLock) TryLock() error {
+	err := kv.RunInNewTxn(l.s, false, func(txn kv.Transaction) error {
+		lockKey := l.getLockKey()
+		val, err := txn.Get([]byte(lockKey))
+		if err != nil {
+			if kv.IsErrNotFound(err) {
+				// if no such lock key, try to lock
+				return l.writeLock(txn)
+			}
+			return errors.Trace(err)
+		}
+
+		// if there's a lock already, check if it's mine
+		var meta LockMeta
+		json.Unmarshal(val, &meta)
+
+		// if another one has the lock, and not timeout yet
+		if meta.ID != l.LockMeta.ID {
+			if oracleTimeNow(txn).Sub(meta.LastUpdated) < 4*leaseInterval {
+				return ErrAlreadyLocked
+			}
+		}
+
+		return l.writeLock(txn)
+	})
+
+	if err == nil {
+		go l.keepalive()
+	}
+	return err
+}
+
+// Lock tries to acquire the lock. it will block until get the lock, or some
+// error happens.
+func (l *DLock) Lock() error {
+	try := 0
+	for {
+		err := l.TryLock()
+		if err == nil {
+			return nil
+		}
+		if terror.ErrorEqual(err, ErrAlreadyLocked) || kv.IsRetryableError(err) {
+			// backoff, take a rest, at least 100ms
+			sleepInterval := 100 + rand.Intn(500)
+			time.Sleep(time.Duration(sleepInterval) * time.Millisecond)
+			try++
+			if try > 10 {
+				log.Warn("retry to fetch lock...", l.name)
+			}
+		} else {
+			log.Error(err)
+			return err
+		}
+	}
+}
+
+// Unlock unlocks the resource.
+func (l *DLock) Unlock() {
+	l.unlockSignal <- struct{}{}
+}
+
+func (l *DLock) getLockKey() []byte {
+	lockKey := fmt.Sprintf("%s-LOCK", l.name)
```

Comment:
name is specified by the user, we can choose a name that doesn't conflict, such as "bootstrap".